neuralop/models/codano.py

```diff
@@ -0,0 +1,371 @@
+            self.cls_token = nn.Parameter(torch.randn(1,
+            hidden_variable_codimension, **self.positional_encoding_modes, dtype=torch.cfloat))
+        
+        if use_positional_encoding:
+            self.positional_encoding = nn.ParameterList()
+            for i in range(self.n_variables):
+                self.positional_encoding.append(
+                    nn.Parameter(
+                        torch.randn(1,
+                            positional_encoding_dim,
+                            *self.positional_encoding_modes,
+                            dtype=torch.cfloat,
+                        )
+                    )
+                )
+    def _extend_positional_encoding(self, num_variables):
+        """
+        Add positional encoding for new variables.
+
+        Parameters
+        ----------
+        num_variables : int
+            number of variables to add
+        """
+        for i in range(num_variables):
+            self.positional_encoding.append(
+                nn.Parameter(
+                    torch.randn(1,
+                        self.positional_encoding_dim,
+                        *self.positional_encoding_modes,
+                        dtype=torch.cfloat,
+                    )
+                )
+            )
+        self.n_variables += num_variables
+
+    def _get_positional_encoding(self, x):
+        encoding_list = []
+        for i in range(self.n_variables):
+            encoding_list.append(torch.fft.irfftn(self.positional_encoding[i],
+                                                s=x.shape[-self.n_dim:]))
+        
+        return torch.stack(encoding_list, dim=1)
+    
+    def _get_cls_token(self, x):
+        cls_token = torch.fft.irfftn(self.cls_token, s=x.shape[-self.n_dim:])
+        repeat_shape = [1 for _ in x.shape]
+        repeat_shape[0] = x.shape[0]
+        cls_token = cls_token.repeat(*repeat_shape)
+        return cls_token
+    
+    def _extend_variables(self, x, static_channel):
+        x = x.unsqueeze(2)
+        if static_channel is not None:
+            repeat_shape = [1 for _ in x.shape]
+            repeat_shape[1] = x.shape[1]
+            static_channel = static_channel.unsqueeze(1).repeat(*repeat_shape)
+            x = torch.cat([x, static_channel], dim=2)
+        if self.use_positional_encoding:
+            positional_encoding = self._get_positional_encoding(x)
+            repeat_shape = [1 for _ in x.shape]
+            repeat_shape[0] = x.shape[0]
+            x = torch.cat([x, positional_encoding.repeat(*repeat_shape)], dim=2)
+        return x
+
+    def forward(self, x: torch.Tensor, static_channel=None):
+        '''
+        Parameters
+        ----------
+        x : torch.Tensor
+            input tensor of shape (batch_size, channel, H, W, ...)
+        static_channel : torch.Tensor
+            static channel tensor of shape (batch_size, static_channel_dim, H, W, ...)
+        '''
+        batch, n_variables, *spatial_shape = x.shape
```

Comment:
is `n_variables` allowed to be `None`? Some explanation of what is expected to happen in that case would be good

Reply:
I renamed the variable and made it different from the 'self.n_variables'.